Tls test

chhsiao90 opened the pull request:
PR for record and discussion
- [x] Test on microproxy.layer.application.tls
- [x] Test on microproxy.protocol.tls

Currently progress: Refactor

@mike820324 , please give me advice if you had time to review!

chhsiao90 committed:
Refactor TlsLayer.

Abstract the TlsLayer logic into two part.
So that it is more readable and testable.

- layer
- protocol.

mike820324 commented on microproxy/utils.py:
```diff
@@ -2,11 +2,18 @@
 import logging.config
 
 
+try:
```
good idea.

chhsiao90 replied:
From mitmproxy, XD
We should run unittest depends on this flag that ignore alpn test if the environment not support alpn.

mike820324 replied:
Agreed.

mike820324 commented:
Would you mind to slightly change your commit into two parts.
- IOStream Cleanup: the iostream.py, which seems like nothing to do with the refactor tls.
- Refactor TLS : rest of the file changed.

Thanks.

chhsiao90 commented:
NP

chhsiao90 committed:
Refactor iostream: though the variable: socket is duplicate with module name: socket, rename it

chhsiao90 committed:
Enrich our StreamClosedError with which type of it

chhsiao90 committed:
Add unittest on TLS related code.

- protocol.tls
- layer.tls

chhsiao90 commented:
TlsLayer done, will start work on protocol.tls module

TravisCI's problem, I will use my linux environment without ALPN support to test on it.
Will fix it later.

chhsiao90 commented:
protocol.tls also done!
The coverage on my machine is 100% for these two module

I will start to pass the test on machine that didn't support alpn

chhsiao90 commented:
Fixed the test! Please help review it, thx.

mike820324 commented on microproxy/test/layer/test_tls.py:
```diff
@@ -0,0 +1,270 @@
+            return_value=self.tls_dest_stream)
+        self.tls_layer.dest_conn = self.dest_conn
+
+    if HAS_ALPN:
```
How bout using `@unittest.skipIf(not HAS_ALPN)`.

mike820324 commented:
Looks really good, I'll may need to take some time to review this. Thanks for the good work.

mike820324 commented on microproxy/protocol/tls.py:
```diff
@@ -91,9 +96,76 @@ def create_src_sslcontext(cert, priv_key, alpn_callback=None,
+        if self.on_alpn:
+            self.on_alpn(conn, alpns)
+        try:
+            return self.alpn_resolver.__call__()
```
why `__call__()` or what's the difference between `self.alpn_resolver()` and `self.apln_resolver.__call__()`

chhsiao90 replied:
The two ways is same.
I'm not sure which way is more readable.

mike820324 commented on microproxy/protocol/tls.py:
```diff
@@ -70,17 +73,19 @@ def certificate_verify_cb(conn, x509, err_num, err_depth, verify_status):
+            certificate_verify_cb)
     else:
-        verify_mode = SSL.VERIFY_NONE
+        def verify_cb_always_true(*args, **kwargs):
```
If I remember correctly there is no need to define a new function here. When the verify mode is `SSL.VERIFY_NONE`, the exception will not raise no matter what the callback function returns when `do_handshake`.

That's why the original code only set the flag, but does not change the callback.

mike820324 committed:
Remove extra function defines.

When verify mode is SSL.VERIFY_NONE,
the SSL.ERROR exception will not be raised whatever the return value is.

chhsiao90 commented on microproxy/protocol/tls.py:
```diff
@@ -74,16 +74,13 @@ def create_dest_sslcontext(insecure=False, trusted_ca_certs="", alpn=None):
-        ssl_ctx.set_verify(SSL.VERIFY_NONE, verify_cb_always_true)
+        verify_mode = SSL.VERIFY_NONE
 
+    ssl_ctx.set_verify(verify_mode, certificate_verify_cb)
```
I'm not sure that the way is better.
It seems more complicated that I need to read to the 83rd line to realize that the verify_mode is used to set the context verify mode.
Not very straightforward for me.
How do you think?

chhsiao90 replied:
What I mean is removing verify_cb_always_true is ok
But no need to create a new variable verify_mode for this.

mike820324 commented:
As for the verify_mode variable, you ate probably right. The variable can be remove, I'll rebase the commit to fix that.

mike820324 commented on microproxy/protocol/tls.py:
```diff
@@ -91,9 +93,76 @@ def create_src_sslcontext(cert, priv_key, alpn_callback=None,
+                  on_alpn=None, alpn_resolver=None):
+        self.alpn_resolver = alpn_resolver
+        self.on_alpn = on_alpn
+        ssl_ctx = create_src_sslcontext(
```
Another question about this part is that why passing two functions instead of just on_alpn? Can you briefly describe the design philosophy behind this part, thanks.
- why passing on_alpn and alpn_resolver
- why not just passing on_alpn to create_src_sslcontext, instead of using self.alpn_callback. 

Thanks

chhsiao90 replied:
For normal case, alpn callback could just return a select alpn with simple logic.

But for our case a proxy server, we need to do a lot of thing when alpn callback,
that need to create connection to destination, and wait for server side handshaking done with selected alpn. It's a lot of things.

The original code is hard to understand that what did alpn callback do, and where and when the alpn decide.
The code I refactor to separate this two logic, that it's simple and easy understand.
When we do handshaking with client side, the on_alpn callback that we register a callback that only do connect to destination
Where and when the alpn selected? We could see that we had a resolver that resolved the destination connection result.

chhsiao90 replied:
But it's still ok to use original alpn_callback, what do you think?

mike820324 replied:
Just saw the message. I was too tired yesterday and fell asleep when I finished my dinner. XD

I think your concern is correct. The original function has done too many things that may be difficult to understand. Separate into two functions is okay for me.

I'll continue reviewing the code, thanks.

mike820324 committed:
Remove verify hostname logic in start_tls in tls protocol.

Since the verification logic is already within MicroProxySSLIOStream,
I remove the verification logic in start_tls in tls protocol.

Moreover, I also modify the MicroProxySSLIOStream to properly raise the
VerificationError when hostname is incorrect.

mike820324 commented:
Force push and rebase to remove the unnecessary verfy_mode variable.

mike820324 commented:
Force push again to squash some commits into one commit entry.

mike820324 commented:
Force push again to merged some commit into one. 
Sorry for the inconvience.
